src/script/standard.cpp

```diff
@@ -44,33 +44,177 @@ const char *GetTxnOutputType(txnouttype t)
+        return CPubKey::ValidSize(pubkey);
+    }
+
+    if (script.size() == CPubKey::COMPRESSED_PUBLIC_KEY_SIZE + 2 && script[0] == CPubKey::COMPRESSED_PUBLIC_KEY_SIZE &&
+        script.back() == OP_CHECKSIG)
+    {
+        pubkey = valtype(script.begin() + 1, script.begin() + CPubKey::COMPRESSED_PUBLIC_KEY_SIZE + 1);
+        return CPubKey::ValidSize(pubkey);
+    }
+    return false;
+}
+
+static bool MatchPayToPubkeyHash(const CScript &script, valtype &pubkeyhash)
+{
+    // Bitcoin address tx, sender provides hash of pubkey, receiver provides signature and pubkey
+    // Template: "OP_DUP << OP_HASH160 << OP_PUBKEYHASH << OP_EQUALVERIFY << OP_CHECKSIG"
+
+    if (script.size() == 25 && script[0] == OP_DUP && script[1] == OP_HASH160 && script[2] == 20 &&
+        script[23] == OP_EQUALVERIFY && script[24] == OP_CHECKSIG)
+    {
+        pubkeyhash = valtype(script.begin() + 3, script.begin() + 23);
+        return true;
+    }
+    return false;
+}
+
+/** Check if a script is of the TX_LABELPUBLIC type
+ *
+ * param script const CScript& a reference to the script to evaluate
+ * param dataCarriage std::vector<valtype> a reference to vector containing
+ *                  2 elements: the big int valu used to mark the txns and the
+ *                  actual string to display along with the txs.
+ *
+ * return boolean
  */
-bool Solver(const CScript &scriptPubKey, txnouttype &typeRet, vector<vector<unsigned char> > &vSolutionsRet)
+static bool MatchLabelPublic(const CScript &script, std::vector<valtype> &dataCarriage)
 {
-    // Templates
-    static multimap<txnouttype, CScript> mTemplates = {
+    // LabelPublc OP_RETURN data size format small
+    // Template: "CScript() << OP_RETURN << OP_BIGINTEGER << OP_DATA"
+    // data deve contenere il OP_DATA
+    if (script.size() < 1 || script[0] != OP_RETURN || !script.IsPushOnly(script.begin() + 1))
+    {
+        return false;
+    }
 
-        // Standard tx, sender provides pubkey, receiver adds signature
-        {TX_PUBKEY, CScript() << OP_PUBKEY << OP_CHECKSIG},
+    valtype data;
+    opcodetype opcode;
+    CScript::const_iterator s = script.begin() + 1;
+    script.GetOp(s, opcode, data);
 
-        // Bitcoin address tx, sender provides hash of pubkey, receiver provides signature and pubkey
-        {TX_PUBKEYHASH, CScript() << OP_DUP << OP_HASH160 << OP_PUBKEYHASH << OP_EQUALVERIFY << OP_CHECKSIG},
+    try
+    {
+        CScriptNum dataId(data, true, 5);
+        dataCarriage.emplace_back(data);
+    }
+    catch (scriptnum_error)
+    {
+        return false;
+    }
+
+    if (script.GetOp(s, opcode, data))
+    {
+        dataCarriage.emplace_back(std::move(data));
+        return true;
+    }
 
-        // Sender provides N pubkeys, receivers provides M signatures
-        {TX_MULTISIG, CScript() << OP_SMALLINTEGER << OP_PUBKEYS << OP_SMALLINTEGER << OP_CHECKMULTISIG},
+    return false;
+}
+
+static bool MatchFreezeCLTV(const CScript &script, std::vector<valtype> &pubkeys)
```

Comment:
Since we no longer need every script to be interpreted via an interface with the same API, why don't you return-in-parameters a CScriptNum and a pubkey, rather then using the single "std::vector<valtype> &pubkeys"

Reply:
I found it more consistent to have all `Match*()` function to have the same interface, but I don't have a strong opinion on this one.